File: Java/src/Distinct Subsequences.java
public class Solution {
    public int numDistinct(String S, String T) {
        if (S.length() < T.length()) return 0;

        int[][] dp = new int[S.length()+1][T.length()+1];

        for (int i = 0; i < dp.length; i++){
            for (int j = 0; j < dp[0].length; j++){
                if (i== 0 && j==0) {
                    dp[i][j] = 1;
                }
                else if (i==0) {
                    dp[i][j] = 0;
                }
                else if (j==0) {
                     dp[i][j] = 1;
                }
                else if (i < j){
                    dp[i][j] = 0;
                }
                else{
                    if (S.charAt(i-1) == T.charAt(j-1)){
                        dp[i][j] = dp[i-1][j-1]+dp[i-1][j];
                    }
                    else{
                         dp[i][j] = dp[i-1][j];
                    }
                }
            }
        }    
        return dp[S.length()][T.length()];
    }
}

File: Python/Distinct Subsequences.py
class Solution:
    # @param S, a string
    # @param T, a string
    # @return an integer
    def numDistinct(self, S, T):
        dp = [[0 for i in range(len(T)+1)] for j in range(len(S)+1)]
        for i in range(0, len(S)+1):
            for j in range(0, len(T)+1):
                if i == 0 and j == 0:
                    dp[i][j] = 1
                elif j == 0:
                    dp[i][j] = 1
                elif i == 0:
                    dp[i][j] = 0
                elif i < j:
                    dp[i][j] = 0
                else:
                    if S[i-1] == T[j-1]:
                        dp[i][j] = dp[i-1][j-1]+dp[i-1][j]
                    else:
                        dp[i][j] = dp[i-1][j]
                print dp
        return dp[len(S)][len(T)]
s = Solution()
print s.numDistinct("AACDE", "ACE")